feat(cli): New Upgrade CLI / Switch to Format Version 3 (upgrade coordination)

Shrekster opened the pull request:
This change lets any Kopia repository owner set an upgrade intent on the repository to enable the draining protocol based upgrade coordination between multiple clients of Kopia and help gain an upgrade quorum before actually upgrading the repository.

- A dedicated upgrade CLI with the ability to set a new upgrade lock, update an existing lock and also revoke a previous orphaned lock.
- The change maintains backwards compatibility with existing `set-parameters --upgrade` based upgrade mechanism. However, it is encouraged to use the new CLI with new projects.
- Adding a new upgrade status section to the Kopia repository status CLI to allow any user to check for an ongoing upgrade status.
- Increments the max supported format version from 2 to 3. Version 3 repository format is not a data format change but has the draining protocol support which helps coordinate an upgrade quorum between all Kopia clients.

https://docs.google.com/document/d/1zIghQc4i4vJSANjSDBOvjzMiqnZSEp1cCasxSUkltSk/edit#heading=h.khston2oxmxy

PRs that introduced the upgrade-lock:

- https://github.com/kopia/kopia/pull/1742
- https://github.com/kopia/kopia/pull/1727
- https://github.com/kopia/kopia/pull/1758

Upgrading from format 1 to 3:
```sh
$ kopia repository upgrade --password 12345678 --config-file /tmp/kopia/repository.config --upgrade-owner-id owner --io-drain-timeout 1s --status-poll-interval 1s --max-clock-drift 1s
Repository upgrade lock intent has been placed.
Waiting for 2s to allow all other kopia clients to drain ...
Waiting for 1s to allow all other kopia clients to drain ...
Waiting for 0s to allow all other kopia clients to drain ...
Successfully drained all repository clients, the lock has been fully-established now.
migrating current indexes to epoch format
Repository indices have been upgraded.
Repository has been successfully upgraded
```

Upgrading from format 3 to 3:
```sh
$ kopia repository upgrade --password 12345678 --config-file /tmp/kopia/repository.config --upgrade-owner-id owner --io-drain-timeout 1s --status-poll-interval 1s --max-clock-drift 1s
ERROR error setting the upgrade lock intent: repository is already upgraded to 3, we can only upgrade till 3
```

Upgrading from format 2 to format 3:
```sh
$ kopia repository upgrade --password 12345678 --config-file /tmp/kopia/repository.config --upgrade-owner-id owner --io-drain-timeout 1s --status-poll-interval 1s --max-clock-drift 1s
Repository upgrade lock intent has been placed.
repository indices have already been migrated to the epoch format
Repository has been successfully upgraded
```

Non-owner upgraders during an upgrade:
```sh
$ kopia repository upgrade --password 12345678 --config-file /tmp/kopia/repository.config --upgrade-owner-id ownerx --io-drain-timeout 10s --status-poll-interval 5s --max-clock-drift 1s
ERROR failed to open repository: repository upgrade in progress
ERROR open repository: unable to open repository: repository upgrade in progress
```

Connections to repository during an upgrade:
```sh
$ kopia repository connect filesystem --password 12345678 --config-file /tmp/kopia/repository.config --path /tmp/kopia/repo/
ERROR failed to open repository: repository upgrade in progress
deleted repository password for /tmp/kopia/repository.config.
kopia: error: error connecting to repository: repository upgrade in progress, try --help
```

Upgrade status with advance notice:
```sh
$ kopia repository status --password 12345678 --config-file /tmp/kopia/repository.config --upgrade-owner-id owner
Config file:         /tmp/kopia/repository.config

Description:         Repository in Filesystem: /tmp/kopia/repo
Hostname:            c02fq0ttmd6r
Username:            shikhar.mall
Read-only:           false
Format blob cache:   15m0s

Storage type:        filesystem
Storage config:      {
                       "path": "/tmp/kopia/repo",
                       "fileMode": 384,
                       "dirMode": 448,
                       "dirShards": [
                         3,
                         3
                       ]
                     }

Unique ID:           83e7ec60180d56181c979213b22b9401c3709d1fcac1a2b2d9d8b921940e0c7e
Hash:                BLAKE2B-256-128
Encryption:          AES256-GCM-HMAC-SHA256
Splitter:            DYNAMIC-4M-BUZHASH
Format version:      3
Content compression: false
Password changes:    false
Max pack length:     20 MiB
Index Format:        v1
Epoch Manager:       disabled
Ongoing upgrade:     Format Version 3 -> 3
Upgrade Time:        2021-12-22 17:05:27.98409 -0800 PST
Upgrade lock:        Unlocked
Lock status:         Draining
```

Upgrade status with immediate lock:
```sh
$ kopia repository status --password 12345678 --config-file /tmp/kopia/repository.config --upgrade-owner-id owner
Config file:         /tmp/kopia/repository.config

Description:         Repository in Filesystem: /tmp/kopia/repo
Hostname:            c02fq0ttmd6r
Username:            shikhar.mall
Read-only:           false
Format blob cache:   15m0s

Storage type:        filesystem
Storage config:      {
                       "path": "/tmp/kopia/repo",
                       "fileMode": 384,
                       "dirMode": 448,
                       "dirShards": [
                         3,
                         3
                       ]                                                                                                                                                                                                                                                                             }

Unique ID:           83e7ec60180d56181c979213b22b9401c3709d1fcac1a2b2d9d8b921940e0c7e
Hash:                BLAKE2B-256-128
Encryption:          AES256-GCM-HMAC-SHA256
Splitter:            DYNAMIC-4M-BUZHASH
Format version:      3
Content compression: false
Password changes:    false
Max pack length:     20 MiB
Index Format:        v1
Epoch Manager:       disabled
Ongoing upgrade:     Format Version 1 -> 3
Upgrade Time:        2021-12-22 16:55:49.98409 -0800 PST
Upgrade lock:        Locked
Lock status:         Fully Established
```

Shrekster commented on cli/command_repository_upgrade.go:
```diff
@@ -0,0 +1,240 @@
+	return func(ctx context.Context, rep repo.DirectRepositoryWriter) error {
+		if !c.skip {
+			if err := act(ctx, rep); err != nil {
+				// exc=plicitly skip all stages on error because tests do not
```
`exc=plicitly` -> `explicitly`

Shrekster replied:
`exc=plicitly` -> `explicitly`

Shrekster commented on cli/command_repository_upgrade.go:
```diff
@@ -0,0 +1,240 @@
+func (c *commandRepositoryUpgrade) setup(svc advancedAppServices, parent commandParent) {
+	cmd := parent.Command("upgrade", "Upgrade repository format.")
+
+	// TODO:
+	// cmd.GetFlag("upgrade-owner-id").Required()
+	// cmd.("owner-id", "The upgrade lock unique owner ID").Required().StringVar(&c.ownerId)
```
I should clean this up.

Shrekster replied:
I should clean this up.

Shrekster commented on cli/command_repository_upgrade_test.go:
```diff
@@ -0,0 +1,205 @@
+	require.Contains(t, out, "Format version:      3")
+}
+
+// TODO(small): add a test to verify that 0.8 client will choke on version-3
```
Let me go ahead an add this as well to this PR.

Shrekster replied:
Let me go ahead an add this as well to this PR.

Shrekster commented on tests/end_to_end_test/suite_test.go:
```diff
@@ -19,3 +19,7 @@ func TestFormatV1(t *testing.T) {
 }
+
+func TestFormatV3(t *testing.T) {
+	testutil.RunAllTestsWithParam(t, &formatSpecificTestSuite{[]string{"--format-version=2"}, content.FormatVersion3})
```
crap, this needs to be `--format-version=3`

Shrekster replied:
crap, this needs to be `--format-version=3`

Shrekster commented:
I will add the over-the-network lock monitor field to the intent as well as this is making the format version change.

Will update again tonight.

adowair commented:
Usage looks good to me. Nice Work! Shall we also the documentation on the [upgrade](https://kopia.io/docs/upgrade/) page?

adowair commented on cli/command_repository_set_parameters_test.go:
```diff
@@ -32,10 +32,13 @@ func (s *formatSpecificTestSuite) TestRepositorySetParameters(t *testing.T) {
 	require.Contains(t, out, "Max pack length:     20 MiB")
 
-	if s.formatVersion == content.FormatVersion1 {
+	switch s.formatVersion {
```
Suggestion: shall we change this to:
```go
require.Contains(t, out, fmt.Sprintf("Format version:      %d", s.formatVersion))
```

Shrekster replied:
will do

adowair commented on cli/command_repository_set_parameters_test.go:
```diff
@@ -111,14 +114,33 @@ func (s *formatSpecificTestSuite) TestRepositorySetParametersUpgrade(t *testing.
+		env.RunAndExpectSuccess(t, "index", "epoch", "list")
+	}
+
+	if s.formatVersion < content.FormatVersion3 {
```
```suggestion
	if s.formatVersion < content.MaxFormatVersion {
```

adowair commented on cli/command_repository_upgrade.go:
```diff
@@ -0,0 +1,240 @@
+	blockUntilDrain bool
+
+	// lock settings
+	advanceNoticeInterval  time.Duration
```
I see we assign`UpgradeLock{AdvanceNoticeDuration: ...}` to this field's value. Should we name this field accordingly? (more changes needed for suggestion).
```suggestion
	advanceNoticeDuration  time.Duration
```

adowair commented on cli/command_repository_upgrade.go:
```diff
@@ -0,0 +1,240 @@
+	cmd.Flag("max-clock-drift", "Maximum tolerated drift on clocks between all Kopia clients").Default("5m").DurationVar(&c.maxPermittedClockDrift)
+	cmd.Flag("block-until-drain", "Drain all the clients but do not perform the upgrade").BoolVar(&c.blockUntilDrain)
+
+	// TODO(small): make rollback a sub-command instead of a flag
```
This looks like it's going to be a small change (I may be wrong)–should we just include ti in this PR?

Shrekster replied:
This ended up becoming a larger change, will look into this in a separate PR.

adowair commented on cli/command_repository_upgrade.go:
```diff
@@ -0,0 +1,240 @@
+		AdvanceNoticeDuration:  c.advanceNoticeInterval,
+		IODrainTimeout:         c.ioDrainTimeout,
+		StatusPollInterval:     c.statusPollInterval,
+		Message:                fmt.Sprintf("Upgrading from format version %d -> %d", mp.Version, content.FormatVersion3),
```
Is this always true?
```suggestion
		Message:                fmt.Sprintf("Upgrading from format version %d -> %d", mp.Version, content.MaxFormatVersion),
```

Shrekster replied:
Yes, that is true as of now.

adowair commented on cli/command_repository_upgrade.go:
```diff
@@ -0,0 +1,240 @@
+	}
+
+	// Update format-blob and clear the cache.
+	// This will fail if we have alread upgraded.
```
```suggestion
	// This will fail if we have already upgraded.
```

adowair commented on cli/command_repository_upgrade_test.go:
```diff
@@ -0,0 +1,205 @@
+			"--max-clock-drift", "1s")
+		require.Contains(t, stderr, "Repository indices have already been migrated to the epoch format, no need to drain other clients")
+		require.Contains(t, stderr, "Repository has been successfully upgraded.")
+	default:
```
Just thinking about the usage here: the "upgrade" API is not idempotent (i.e. two subsequent upgrade commands won't give the same result). I'm sure if we publish documentation it'll be okay, but I wonder if there's any potential for trouble.

adowair reviewed (commented):
Thanks for putting this up, some minor nits and questions

jkowalski commented on cli/app.go:
```diff
@@ -231,6 +233,8 @@ func (c *App) setup(app *kingpin.Application) {
 	app.Flag("disable-internal-log", "Disable internal log").Hidden().Envar("KOPIA_DISABLE_INTERNAL_LOG").BoolVar(&c.disableInternalLog)
 	app.Flag("advanced-commands", "Enable advanced (and potentially dangerous) commands.").Hidden().Envar("KOPIA_ADVANCED_COMMANDS").StringVar(&c.AdvancedCommands)
+	app.Flag("upgrade-owner-id", "Repository format upgrade owner-id.").Hidden().Envar("KOPIA_REPO_UPGRADE_OWNER_ID").StringVar(&c.upgradeOwnerID)
+	app.Flag("upgrade-no-block", "Do not block when repository format upgrade is in progress.").Default("false").Envar("KOPIA_REPO_UPGRADE_NO_BLOCK").BoolVar(&c.doNotWaitForUpgrade)
```
that should be hidden

Shrekster commented:
Moving the upgrade lock under maintenance schedule blob. Will update again shorty and put it back in review.

Shrekster commented:
@julio-lopez PTAL

julio-lopez commented on cli/app.go:
```diff
@@ -243,6 +245,8 @@ func (c *App) setup(app *kingpin.Application) {
 	app.Flag("advanced-commands", "Enable advanced (and potentially dangerous) commands.").Hidden().Envar(c.EnvName("KOPIA_ADVANCED_COMMANDS")).StringVar(&c.AdvancedCommands)
 	app.Flag("track-releasable", "Enable tracking of releasable resources.").Hidden().Envar(c.EnvName("KOPIA_TRACK_RELEASABLE")).StringsVar(&c.trackReleasable)
 	app.Flag("dump-allocator-stats", "Dump allocator stats at the end of execution.").Hidden().Envar(c.EnvName("KOPIA_DUMP_ALLOCATOR_STATS")).BoolVar(&c.dumpAllocatorStats)
+	app.Flag("upgrade-owner-id", "Repository format upgrade owner-id.").Hidden().Envar(c.EnvName("KOPIA_REPO_UPGRADE_OWNER_ID")).StringVar(&c.upgradeOwnerID)
```
As discussed offline: let's use the repo (maintenance) owner, instead of introducing another entity here. This means, the upgrade needs to be done from the repo owner instance. It has the advantage that 1. ensures maintenance is not running concurrently, and 2. it allows delayed execution (as part of maintenance) if we decide to do so in the future.

Shrekster replied:
Discussed with @julio-lopez , will make this change in a separate PR.

julio-lopez commented on cli/app.go:
```diff
@@ -243,6 +245,8 @@ func (c *App) setup(app *kingpin.Application) {
 	app.Flag("track-releasable", "Enable tracking of releasable resources.").Hidden().Envar(c.EnvName("KOPIA_TRACK_RELEASABLE")).StringsVar(&c.trackReleasable)
 	app.Flag("dump-allocator-stats", "Dump allocator stats at the end of execution.").Hidden().Envar(c.EnvName("KOPIA_DUMP_ALLOCATOR_STATS")).BoolVar(&c.dumpAllocatorStats)
+	app.Flag("upgrade-owner-id", "Repository format upgrade owner-id.").Hidden().Envar(c.EnvName("KOPIA_REPO_UPGRADE_OWNER_ID")).StringVar(&c.upgradeOwnerID)
+	app.Flag("upgrade-no-block", "Do not block when repository format upgrade is in progress.").Hidden().Default("false").Envar(c.EnvName("KOPIA_REPO_UPGRADE_NO_BLOCK")).BoolVar(&c.doNotWaitForUpgrade)
```
do these flags need to be "global"? can they be made specific to the upgrade command?

Shrekster replied:
these need to be supplied when you are connecting to the repo, so yes global

julio-lopez commented on cli/app.go:
```diff
@@ -126,7 +126,9 @@ type App struct {
 
-	observability observabilityFlags
+	observability       observabilityFlags
+	upgradeOwnerID      string
```
see comment below about _upgrade owner_

julio-lopez commented on cli/command_repository_set_parameters_test.go:
```diff
@@ -127,14 +123,35 @@ func (s *formatSpecificTestSuite) TestRepositorySetParametersUpgrade(t *testing.
+			"--upgrade-owner-id", "owner",
+			"--io-drain-timeout", "1s", "--force",
+			"--status-poll-interval", "1s",
+			"--max-clock-drift", "1s")
```
What's the difference between the `if` and `else` blocks?

Shrekster replied:
`RunAndExpectFailure` vs `RunAndExpectSuccess`

But I can still make this a bit nicer.

Shrekster replied:
refactored the code slightly

https://github.com/kopia/kopia/pull/1818/commits/4f4917ec02093d3d222205ad5b3348477f519ca6

julio-lopez reviewed (commented):
preliminary comments / questions

Shrekster commented:
@julio-lopez @adowair PTAL

- dropped --max-clock-drift flag, converting that into a const (still allowing tests to override so they do not have to block for the long default wait)
- dropped --block-until-drain flag, this was used for advance-notice feature. We can add this later as needed.
- updated documentation on --upgrade-no-block flag as it was used by status cli.
- Added subcommands to upgrade cli: begin & rollback.

julio-lopez commented on cli/command_repository_upgrade.go:
```diff
@@ -0,0 +1,291 @@
+	// upgrade phases
+
+	// Set the upgrade lock intent.
+	beginCmd.Action(svc.directRepositoryWriteAction(c.runPhase(c.setLockIntent)))
+	// If requested then drain all the clients otherwise stop here.
+	beginCmd.Action(svc.directRepositoryWriteAction(c.runPhase(c.drainOrCommit)))
+	// If the lock is fully established then perform the upgrade.
+	beginCmd.Action(svc.directRepositoryWriteAction(c.runPhase(c.upgrade)))
+	// Commit the upgrade and revoke the lock, this will also cleanup any
+	// backups used for rollback.
+	beginCmd.Action(svc.directRepositoryWriteAction(c.runPhase(c.commitUpgrade)))
```
Why are these separate command `Action`s, more specifically, why are thy separate repo sessions (`directRepositoryWriteAction`)? Do they need to be?

Shrekster replied:
Good question, each of these functions are individual idempotent components which are retryable. Also, it is easier to read it in phases of intent -> drain -> upgrade -> commit.

We do not need this to be separate write sessions though, I can fix this in a follow-up PR.

julio-lopez commented on cli/command_repository_upgrade.go:
```diff
@@ -0,0 +1,291 @@
+}
+
+const (
+	experimentalWarning = `WARNING: The upgrade command is an EXPERIMENTAL feature. Please use with caution.
```
```suggestion
	experimentalWarning = `WARNING: The upgrade command is an EXPERIMENTAL feature. Please DO NOT use it, it may corrupt your repository and cause data loss.
```

julio-lopez commented on cli/command_repository_upgrade.go:
```diff
@@ -0,0 +1,291 @@
+			return errors.Wrap(err, "failed to get upgrade lock intent")
+		}
+
+		if l.AdvanceNoticeDuration == 0 {
```
I do not understand the behavior here. If no (advance) notice was set, then no client draining is performed? that seems incorrect, unless I'm missing something here.

Shrekster replied:
the if conditions wrapping this and the the advance notice block here is saying this:

if we have already upgraded (to version 2) and no advance notice was set then let the upgrade continue

julio-lopez commented on cli/command_repository_upgrade.go:
```diff
@@ -0,0 +1,291 @@
+			return errors.Wrap(err, "upgrade lock got revoked after the intent was placed, giving up")
+		}
+
+		// TODO: this can get stuck
```
why? how?

Shrekster replied:
on Macs and systems which have resume from sleep capability, certain time jumps could cause a longer than expected wait time here

julio-lopez commented on cli/command_repository_upgrade.go:
```diff
@@ -0,0 +1,291 @@
+
+func (c *commandRepositoryUpgrade) runPhase(act func(context.Context, repo.DirectRepositoryWriter) error) func(context.Context, repo.DirectRepositoryWriter) error {
+	return func(ctx context.Context, rep repo.DirectRepositoryWriter) error {
+		if !c.skip {
+			if err := act(ctx, rep); err != nil {
+				// Explicitly skip all stages on error because tests do not
+				// skip/exit on error. Tests override os.Exit() that prevents
+				// running rest of the phases until we set the skip flag here.
+				// This flag is designed for testability and also to support
+				// rollback.
+				c.skip = true
+				return err
+			}
+		}
+
+		return nil
```
Invert the check here to simplify these blocks?

```suggestion
		if c.skip {
			return nil
		}
		if err := act(ctx, rep); err != nil {
			// Explicitly skip all stages on error because tests do not
			// skip/exit on error. Tests override os.Exit() that prevents
			// running rest of the phases until we set the skip flag here.
			// This flag is designed for testability and also to support
			// rollback.
			c.skip = true
			return err
		}

```

Shrekster replied:
thank you! committed

julio-lopez commented on cli/command_repository_upgrade.go:
```diff
@@ -0,0 +1,291 @@
+	beginCmd := parent.Command("begin", "Begin upgrade.").Default()
+	beginCmd.Flag("advance-notice", "Advance notice for upgrade to allow enough time for other Kopia clients to notice the lock").DurationVar(&c.advanceNoticeDuration)
+	beginCmd.Flag("io-drain-timeout", "Max time it should take all other Kopia clients to drop repository connections").Default(repo.DefaultRepositoryBlobCacheDuration.String()).DurationVar(&c.ioDrainTimeout)
+	beginCmd.Flag("force", "Force using an unsafe io-drain-timeout").Default("false").Hidden().BoolVar(&c.force)
```
```suggestion
	beginCmd.Flag("allow-unsafe-upgrade", "Force using an unsafe io-drain-timeout for the upgrade").Default("false").Hidden().BoolVar(&c.force)
```

julio-lopez commented on cli/command_repository_upgrade.go:
```diff
@@ -0,0 +1,293 @@
+// skipped until the lock is fully established.
+func (c *commandRepositoryUpgrade) drainOrCommit(ctx context.Context, rep repo.DirectRepositoryWriter) error {
+	cf := rep.ContentReader().ContentFormat()
+	if cf.MutableParameters.EpochParameters.Enabled {
```
Is it possible to have a repository in version 2 with `EpochParameters.Enabled == false`?

julio-lopez commented on cli/command_repository_upgrade.go:
```diff
@@ -0,0 +1,293 @@
+		log(ctx).Infof("Waiting for %s to allow all other kopia clients to drain ...", upgradeTime.Sub(rep.Time()).Round(time.Second))
+
+		locked, writersDrained := l.IsLocked(now)
+		if locked {
+			if writersDrained {
+				// we have the lock now
+				break
+			}
+		} else {
+			return errors.Wrap(err, "upgrade lock got revoked after the intent was placed, giving up")
+		}
```
```suggestion
		if !locked {
			return errors.Wrap(err, "upgrade lock got revoked after the intent was placed, giving up")
		}

		if writersDrained {
			// we have the lock now
			break
		}
```
